src/checkers/inference/solver/backend/encoder/binary/ComparableConstraintEncoder.java

```diff
@@ -0,0 +1,7 @@
+package checkers.inference.solver.backend.encoder.binary;
+
+/**
+ * Created by mier on 07/11/17.
+ */
+public interface ComparableConstraintEncoder<ConstraintEncodingT> extends BinaryConstraintEncoder<ConstraintEncodingT> {
```

Comment:
One question: What is the benefit of having those empty interfaces for each different constraint type? There seems no difference to let concrete ConstraintEncoder directly implements `BinaryConstraintEncoder` or `CombineConstraintEncoder`, because this interface is empty and add no new APIs.

And also, for a new comer, when he/she looks a concrete ConstraintEncoder implementation, and trace up to one of those empty interfaces, it seems a bit confusing to him/her....

Can we simply remove those empty interfaces?

Reply:
Interfaces that have empty methods are called marker interfaces. Its purpose is to tell compiler statically about the information about the instances of that interface.
Truly, Subtype, Equality, Inequality and ComparableConstraintEncoder interfaces doesn't define any new methods, but it successfully restricted what kind of instances when you override the factory methods to create ConstraintEncoders.
For example, we have two methods in AbstractFormatTranslator:
`    protected SubtypeConstraintEncoder<ConstraintEncodingT> createSubtypeConstraintEncoder(ConstraintVerifier verifier) {
        return null;
    }`
`    protected EqualityConstraintEncoder<ConstraintEncodingT> createEqualityConstraintEncoder(ConstraintVerifier verifier) {
        return null;
    }`
If you don't have that marker interfaces, you'll use BinaryConstraintEncoder as the return type for both methods. As a result, you can either pass "EqualityConstraintEncoder" in createSubtypeConstaintEncoder and mess up the four encoders in each other by accident. But if you have finer granularity using the marker interface, you are only allowed to pass the subtypeconstrainencoder in createSubtypeConstraintEncoder, but not in createEqualityConstraintEncoder, or the other two.

Reply:
Marker interfaces usually used as a mark, so that some place will have a check on the mark and have some special process with that mark, e.g. a class have to implement  empty interface `Cloneable` to mark this class is cloneable. (Some place will have an `obj instance of Cloneable` check, and throws Exception if that `obj` does not implement `Cloneable`.

However, here these marker interfaces are used as Type Restrictions --- i.e. there were no special place to check those marks except subtype checking. Then it is actually no very helpful, as we gain no better safety for those marker interfaces. Let's illustrate in your use case: here you try to avoid sub-class passed wrong encoders for each factory method. Your argument is by having those marker interfaces, we can avoid passing a `EqualityEncoder` to a `createSubtypeEncoder` method. This is true. However, why a programmer could make this mistake?( Because he can see the method name is `createSubtypeEncoder`, and why does he will return a `EqualityEncoder`?)

Ok, let's just assume this kind of mistake might happen. Then, a similar mistake would be having a `xxxSubtypeEncoder implements SubtypeEncoder` by its actual implementation is about encoding `EqualityConstriant` (This is statically allowed, as the parent interface methods takes in a `BinaryConstraint`).

Then, your marker interface cannot prevent this similar mistake.

Why these two mistake are similar? As programmer provide a wrong implementation, even with given a clearly text information about what he should implement (i.e. return `EqualityEncoder` in `createSubtypeEncoder`, or implement `EqualityConstraint` encoding in a `SubtypeEncoder`).

So my conclusion is, let's trust programmer will provide correct implementation with a given clearly defined text information. As these marker interfaces actually doesn't prevent this mistake at the root.

```

Reply:
> " have some special process with that mark".

`instance of Cloneable` check at runtime doesn't have black-box logic to test if an object is really clonable or not. It only "shallowly" checks an object is implementing Clonable interface or not. If a call to clone() method sees an object implementing the interface, it will pass the call. That special processing is now static. If a class doesn't implement correct encoder, then it is rejected.
 
> However, why a programmer could make this mistake?

By accident. I have experience that I mistyped something because of IDE auto completion. Sometimes it happens.

Marker interface is also a way to group together classes with similar functionality. Instead of putting all MaxSATSubtype, MaxSATEquality,  MaxSATInequality,  MaxSATComparable and the other eight all under BinaryConstraintEncoder, I like the feeling that they are grouped together by an interface that differentiate them. May I'm curious which solver backend doesn't support traslating preference constraint. If we have that marker interface, we can simply check who doesn't implement the PreferenceConstraintEncoder. It's crazily easier to look for it from at most three entries(each backend has one) than looking through at most 12 entries. If by accident, programmers are implementing wrong interface, e.g. EqualityEncoder implementing SubtypeEncoder interface, then it's also very easy to find it out when I investigate who implements SubtypeEncoder interface. But using only BinaryEncoder interface, I miss the second opportunity to catch this bug.
So overall, I like we have more control over what we're doing conceptually. Not just programmingly.

Reply:
This is a screen shot : each class is listed clearly under its category.
![screenshot from 2017-11-09 10-35-12](https://user-images.githubusercontent.com/15336177/32613779-c474ceba-c539-11e7-954f-dd3418e4ccde.png)

Reply:
Um... Grouping makes sense. Sure, having those markers interfaces do help us to group implementations. Thanks for the discussion!

Reply:
Also thank you!